lib/DXIL/DxilResourceProperties.cpp

```diff
@@ -0,0 +1,248 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// DxilResourceProperites.cpp                                                //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "dxc/DXIL/DxilResourceProperties.h"
+#include "llvm/IR/Constant.h"
+#include "dxc/DXIL/DxilShaderModel.h"
+#include "llvm/IR/DerivedTypes.h"
+#include "llvm/IR/Constants.h"
+#include "dxc/DXIL/DxilResourceBase.h"
+#include "dxc/DXIL/DxilResource.h"
+#include "dxc/DXIL/DxilCBuffer.h"
+#include "dxc/DXIL/DxilSampler.h"
+#include "dxc/DXIL/DxilOperations.h"
+#include "dxc/DXIL/DxilInstructions.h"
+
+using namespace llvm;
+
+namespace hlsl {
+
+bool DxilResourceProperties::operator==(const DxilResourceProperties &RP) {
+  return Class == RP.Class && Kind == RP.Kind && RawDword0 == RP.RawDword0 &&
+         RawDword1 == RP.RawDword1;
+}
+
+bool DxilResourceProperties::operator!=(const DxilResourceProperties &RP) {
+  return !(*this == RP) ;
+}
+
+namespace resource_helper {
+// Resource Class and Resource Kind is used as seperate parameter, other fileds
+// are saved in constant.
+// The constant is as struct with int32 fields.
+// ShaderModel 6.6 has 2 fileds.
+Constant *getAsConstant(const DxilResourceProperties &RP, Type *Ty,
+                        const ShaderModel &) {
```

Comment:
DXASSERT SM 6.6 to ensure code is updated if shader model is updated which may have more resource properties.